core/icons/mutator_icon.ts

```diff
@@ -0,0 +1,358 @@
+  private createRootBlock() {
+    this.rootBlock = this.sourceBlock.decompose!(
+      this.miniWorkspaceBubble!.getWorkspace()
+    )!;
+
+    for (const child of this.rootBlock.getDescendants(false)) {
+      child.queueRender();
+    }
+
+    this.rootBlock.setMovable(false);
+    this.rootBlock.setDeletable(false);
+
+    const flyoutWidth =
+      this.miniWorkspaceBubble?.getWorkspace()?.getFlyout()?.getWidth() ?? 0;
+    this.rootBlock.moveBy(
+      this.rootBlock.RTL ? -(flyoutWidth + this.MARGIN) : this.MARGIN,
+      this.MARGIN
+    );
+  }
+
+  /** Adds a listen to the source block that triggers saving connections. */
+  private addSaveConnectionsListener() {
+    if (!this.sourceBlock.saveConnections || !this.rootBlock) return;
+    const saveConnectionsListener = () => {
+      if (!this.sourceBlock.saveConnections || !this.rootBlock) return;
+      this.sourceBlock.saveConnections(this.rootBlock);
+    };
+    saveConnectionsListener();
+    this.sourceBlock.workspace.addChangeListener(saveConnectionsListener);
+  }
+
+  /**
+   * Creates a change listener to add to the mini workspace which recomposes
+   * the block.
+   */
+  private createMiniWorkspaceChangeListener() {
+    return (e: Abstract) => {
+      if (!this.shouldIgnoreMutatorEvent(e) && !this.updateWorkspacePid) {
+        this.updateWorkspacePid = setTimeout(() => {
+          this.updateWorkspacePid = null;
+          this.recomposeSourceBlock();
+        }, 0);
+      }
+    };
+  }
+
+  /**
+   * Returns true if the given event is not one the mutator needs to
+   * care about.
+   */
+  private shouldIgnoreMutatorEvent(e: Abstract) {
+    return (
+      e.isUiEvent ||
+      e.type === eventUtils.CREATE ||
+      (e.type === eventUtils.CHANGE &&
+        (e as BlockChange).element === 'disabled')
+    );
+  }
+
+  /** Recomposes the source block based on changes to the mini workspace. */
+  private recomposeSourceBlock() {
+    if (!this.rootBlock) return;
+
+    const existingGroup = eventUtils.getGroup();
+    if (!existingGroup) eventUtils.setGroup(true);
+
+    const oldExtraState = BlockChange.getExtraBlockState_(this.sourceBlock);
+    this.sourceBlock.compose!(this.rootBlock);
+    const newExtraState = BlockChange.getExtraBlockState_(this.sourceBlock);
+
+    if (oldExtraState !== newExtraState) {
+      eventUtils.fire(
+        new (eventUtils.get(eventUtils.BLOCK_CHANGE))(
+          this.sourceBlock,
+          'mutation',
+          null,
+          oldExtraState,
+          newExtraState
+        )
+      );
+    }
+
+    eventUtils.setGroup(existingGroup);
+  }
+
+  /** @internal */
+  getWorkspace(): WorkspaceSvg | undefined {
+    return this.miniWorkspaceBubble?.getWorkspace();
+  }
+
+  /**
+   * Reconnects the given connection to the mutated input on the given block.
+   */
+  static reconnect(
+    connectionChild: Connection,
+    block: Block,
+    inputName: string
+  ): boolean {
```

Comment:
Please provide JSDocs for parameters and return value.

(Actually, for all the methods in this class, but especially here were the meaning of the return value is not obvious.)

Reply:
Also: my usual rule of thumb is that I only make a function `f` a static method on class `C` if it either takes one or more `C` instances as parameters, or returns a `C` instance, or otherwise at least mentions `C` somewhere in its body.  But this looks more like a random helper function that could be exported on its own (if indeed it is exported at all, though that might be useful for subclassing.)

Reply:
Yeah I have no idea where this method should live. It is just a random helper method which is usually used with mutators.

I don't want to export it on `MutatorIcon` or `Blockly.icons` but idk where else to put it.